src/lib/FreeC/Backend/Coq/Converter/Expr.hs

```diff
@@ -230,6 +230,18 @@ convertExpr' (IR.Lambda _ args expr _) [] [] = localEnv $ do
   expr' <- convertExpr expr
   foldrM (generatePure .: Coq.Fun . return) expr' args'
 
+convertExpr' (IR.Let _ binds expr _) [] [] = localEnv $ do
+  let bindPatsExprs = map (\x -> (IR.bindVarPat x, IR.bindExpr x)) binds
+  expr' <- convertExpr expr -- wrong order ?
+  foldrM helper expr' bindPatsExprs
+   where
+    helper :: (IR.VarPat, IR.Expr) -> Coq.Term -> Converter Coq.Term
+    helper (p, e) inExprTerm = do
+      e'      <- convertExpr e
+      qualid  <- renameAndDefineVar (IR.varPatSrcSpan p) (IR.varPatIsStrict p) (IR.varPatIdent p) (IR.varPatType p)
+      varPatType' <- mapM convertType (IR.varPatType p)
+      return (Coq.Let qualid [] varPatType' e' inExprTerm)
```

Comment:
The line marked with `--wrong order ?` is probably causing the "Identifier not in scope" error. `convertExpr` has to look up `b` but `b` is inserted by `helper` into the environment. With `foldr` instead of `foldrM` it should be possible to fix the order.

```suggestion
convertExpr' (IR.Let _ binds expr _) [] [] = localEnv $ do
  let bindPatsExprs = map (\x -> (IR.bindVarPat x, IR.bindExpr x)) binds
  foldr helper (convertExpr expr) bindPatsExprs
   where
    helper :: (IR.VarPat, IR.Expr) -> Converter Coq.Term -> Converter Coq.Term
    helper (p, e) inExprConverter = do
      e'      <- convertExpr e
      qualid  <- renameAndDefineVar (IR.varPatSrcSpan p) (IR.varPatIsStrict p) (IR.varPatIdent p) (IR.varPatType p)
      varPatType' <- mapM convertType (IR.varPatType p)
      Coq.Let qualid [] varPatType' e' <$> inExprConverter
```

Reply:
That is a really nice solution i did not think of. And it fixes the bug. Thank you!